### DISCUSSION TITLE
User+password based authentication no longer works for submodule updates (since 36.10.0)

### DISCUSSION THREAD

**Discussion by lcobucci** [renovatebot/renovate#24077] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Latest (36.58.0)

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

Since v36.10.0, Renovate is creating a URL that only contains a token (instead of username+pwd), leading to authentication errors as the password can't be read.

I've found two patches that seem to be related:

- https://github.com/renovatebot/renovate/pull/23458
- https://github.com/renovatebot/renovate/pull/22061

Although they seem like a very good improvement, they are only considering token-based authentication.

We're ingesting host configuration via environment variables (hostname redacted):

- `GIT__REFS_MYGITLABHOST_COM_USERNAME`
- `GIT__REFS_MYGITLABHOST_COM_PASSWORD`

Once we pinned our runners to use v36.9.2, everything was working fine again :smile:

Looking at the test suite, there's no test covering that type of authentication.
So, something like the code below should reproduce the issue:

```js

    it('combined username+pwd from host rule is used to detect branch', async () => {
      gitMock.listRemote.mockResolvedValueOnce(
        'ref: refs/heads/main HEAD\n5701164b9f5edba1f6ca114c491a564ffb55a964        HEAD'
      );
      hostRules.add({
        hostType: 'github',
        matchHost: 'github.com',
        username: 'x-access-token',
        password: 'ghs_abc123',
      });
      const res = await extractPackageFile('', '.gitmodules.2', {});
      expect(res?.deps).toHaveLength(1);
      expect(res?.deps[0].currentValue).toBe('main');
      expect(gitMock.env).toHaveBeenCalledWith({
        GIT_CONFIG_COUNT: '3',
        GIT_CONFIG_KEY_0:
          'url.https://x-access-token:ghs_abc123@github.com/.insteadOf',
        GIT_CONFIG_KEY_1:
          'url.https://x-access-token:ghs_abc123@github.com/.insteadOf',
        GIT_CONFIG_KEY_2:
          'url.https://x-access-token:ghs_abc123@github.com/.insteadOf',
        GIT_CONFIG_VALUE_0: 'ssh://git@github.com/',
        GIT_CONFIG_VALUE_1: 'git@github.com:',
        GIT_CONFIG_VALUE_2: 'https://github.com/',
      });
      expect(gitMock.listRemote).toHaveBeenCalledWith([
        '--symref',
        'https://github.com/PowerShell/PowerShell-Docs',
        'HEAD',
      ]);
    });
```

### Relevant debug logs

<details><summary>Logs</summary>

```
(...)

DEBUG: Adding password authentication for mygitlabhost.com to hostRules

(...)

WARN: Error mapping git submodules during extraction (repository=my-repo-name, packageFile=.gitmodules)
       "err": {
         "task": {
           "commands": [
             "ls-remote",
             "--symref",
             "https://**redacted**@mygitlabhost.com/tools/submodule.git",
             "HEAD"
           ],
           "format": "utf-8",
           "parser": "[function]"
         },
         "message": "fatal: could not read Password for '[https://**redacted**@mygitlabhost.com'](https://**redacted**@mygitlabhost.com%27/): No such device or address\n",
         "stack": "Error: fatal: could not read Password for '[https://**redacted**@mygitlabhost.com'](https://**redacted**@mygitlabhost.com%27/): No such device or address\n\n    at Object.action (/usr/local/lib/node_modules/renovate/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/usr/local/lib/node_modules/renovate/node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)\n    at /usr/local/lib/node_modules/renovate/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:127:42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/usr/local/lib/node_modules/renovate/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)\n    at GitExecutorChain.<anonymous> (/usr/local/lib/node_modules/renovate/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/usr/local/lib/node_modules/renovate/node_modules/simple-git/dist/cjs/index.js:55:24)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)"
       }
(...)
```
</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
@Shegox is this a feature you're interested in adding (restoring?)

**Reply by Shegox** [reply]
Can do, I probably have sometime today afternoon to look at it. 

I will investigate and can probably open a pull request for this.

**Reply by Shegox** [reply]
Created https://github.com/renovatebot/renovate/pull/24081

**Reply by lcobucci** [reply]
#24081 didn't address the issue as other components seem to be affecting the URL used for the `ls-remote` command.
More info: https://github.com/renovatebot/renovate/pull/24081#issuecomment-1705439132

**Reply by Shegox** [reply]
Moving the comment over:

> Okay, after adding a whole lot of `console.log()` on a few places I've managed to find the issue... it's a tricky one, I think :sob: 
> 
> We were previously relying on the host types `git-refs` and `git-tags` to resolve the submodules.
> The authentication changes aren't considering those host types, leading the creation of environment variables using the token as auth mechanism (`RENOVATE_TOKEN` is automatically promoted as a host rule).
> 
> So, essentially I have the following host rules:
> 
> ```js
> [
>   {
>     hostType: 'github',
>     matchHost: 'github.com',
>     token: 'github-token',
>     resolvedHost: 'github.com'
>   },
>   {
>     hostType: 'git-refs',
>     matchHost: 'mygitlab.com',
>     password: 'mygitlab-pwd',
>     username: 'renovate-bot',
>     resolvedHost: 'mygitlab.com'
>   },
>   {
>     hostType: 'git-tags',
>     matchHost: 'mygitlab.com',
>     password: 'mygitlab-pwd',
>     username: 'renovate-bot',
>     resolvedHost: 'mygitlab.com'
>   },
>   {
>     matchHost: 'mygitlab.com',
>     token: 'mygitlab-pwd',
>     hostType: 'gitlab',
>     resolvedHost: 'mygitlab.com'
>   }
> ]
> ```
> 
> The host rule created from `RENOVATE_TOKEN` is legit for the API requests. However, not for the Git related requests.
> 
> Filtering for `git-refs`, `renovate` creates the following env vars (contents of `newEnvironmentVariables`):
> 
> ```js
> {
>   GIT_CONFIG_KEY_0: 'url.https://ssh:github-token@github.com/.insteadOf',
>   GIT_CONFIG_VALUE_0: 'ssh://git@github.com/',
>   GIT_CONFIG_KEY_1: 'url.https://git:github-token@github.com/.insteadOf',
>   GIT_CONFIG_VALUE_1: 'git@github.com:',
>   GIT_CONFIG_KEY_2: 'url.https://github-token@github.com/.insteadOf',
>   GIT_CONFIG_VALUE_2: 'https://github.com/',
>   GIT_CONFIG_COUNT: '6',
>   GIT_CONFIG_KEY_3: 'url.https://renovate-bot:mygitlab-pwd@mygitlab.com/.insteadOf',
>   GIT_CONFIG_VALUE_3: 'ssh://git@mygitlab.com/',
>   GIT_CONFIG_KEY_4: 'url.https://renovate-bot:mygitlab-pwd@mygitlab.com/.insteadOf',
>   GIT_CONFIG_VALUE_4: 'git@mygitlab.com:',
>   GIT_CONFIG_KEY_5: 'url.https://renovate-bot:mygitlab-pwd@mygitlab.com/.insteadOf',
>   GIT_CONFIG_VALUE_5: 'https://mygitlab.com/'
> }
> ```
> And manually running `git ls-remote --symref https://mygitlab.com/path/to/my/repo.git HEAD` with those variables work just fine. However, letting the command run leads to the password prompt of a completely weird username (`oauth2`) - which understandably fails on non-interactive sessions.
> 
> I also confirm that running `git ls-remote --symref https://oauth2@mygitlab.com/path/to/my/repo.git HEAD` with those variables **doesn't work at all** (which makes sense because there's no match on `https://oauth2@...`.
> 
> What do you think? We might still be missing some scenarios for `getAuthenticationRules()` or `getGitEnvironmentVariables()` - it seems like different components are messing up with the expectations here.
> 
> _Originally posted by @lcobucci in https://github.com/renovatebot/renovate/issues/24081#issuecomment-1705439132_
>             

If I understand it correctly, the main problem seems to be caused that you have two hostRules for the same gitlab-server, right? And Renovate picks the only the token one (the wrong one)?

I can see if I can get a GitLab server to fully reproduce your problem and then see if we can further improve our handling logic.

**Reply by lcobucci** [reply]
@Shegox thanks for bringing the context here.

To summarise the wall of text: the previously working host rules are now ignored and other parts of renovate are overwriting the URL used to retrieve the commit hash of the submodule. "oauth2" is being added as user, which doesn't align with the expectations of the test scenarios for `git/auth.ts`.

**Reply by Shegox** [reply]
@lcobucci I assume you have a local (development) version running of Renovate, right? 

Can you please try one thing for me:
Can you try once with a hostrule like this `hostType = gitlab` instead of `git-refs`? And let me know if that causes any difference? Background is that the `git-submodules` manager currently only adds hostRules for known git hosts and doesn't add hostRules for `git-refs/tags` when extracting the dependency or updating the submodule.
   ```json
   [
     {
       "hostType": "gitlab",
       "matchHost": "mygitlab.com",
       "password": "mygitlab-pwd",
       "username": "renovate-bot"
     }
   ] 
   ```

**Reply by Shegox** [reply]
I got now a Gitlab instance and are able to reproduce it. 
I have two "repository tokens", one for the repository I'm renovating as well as another one for the "submodule".

Some observations: 
- Following hostRule works: 
  ```js
  {
    hostType: "gitlab",
    matchHost: "gitlab.int",
    token: "tokenForSubmodule"
  },
  ```
    and omitting the hostType as well:
  ```js
  {
    matchHost: "gitlab.int",
    token: "tokenForSubmodule"
  },
  ```
- The same host rule with username password doesn't work:
    ```js
    {
      hostType: "gitlab",
      matchHost: "gitlab.int",
      username: "oauth2",
      password: "tokenForSubmodule"
    },
  ```
  This hostrule causes Renovate to completely fail the cloning of the repository. I assume because the username/password is then used to try to interact with the actual repository.
  
It seems that a code change to `getGitEnvironmentVariables(['git-tags', git-refs']);` for git-submodules allows for Renovate to pick up additional hostrules matching that host and will perform the correct lookup without additional negative side-affects and allows for the configuration of a hostrules like: 
```js
    {
      hostType: "git-refs",
      matchHost: "gitlab.int",
      username: "oauth2",
      password: "tokenForSubmodule"
    },
    {
      hostType: "git-tags",
      matchHost: "gitlab.int",
      username: "oauth2",
      password: "tokenForSubmodule"
    },
```

I think adding the support for `git-refs/git-tags` to the supported hostrules for `git-submodules` is reasonable. @rarkins what do you think?

If you want you can already try the draft PR: #24250

**Reply by lcobucci** [reply]
@Shegox I'll do some testing tomorrow morning - today is insanely busy.

**Reply by lcobucci** [reply]
@Shegox sorry about the delay.
I've tested your PR and it didn't work.

Reverting the changes and adding some good old `console.log()` calls, I've managed to find that on:

https://github.com/renovatebot/renovate/blob/4b616c29c31b3cf891c1244fabae944cea9efb65/lib/modules/manager/git-submodules/extract.ts#L120-L121

1. `subModuleUrl` has the full authentication string (already using HTTP, as I'm using relative URL for the submodule): `https://oauth2:my-secret-token@mygitlab.com/path/to/my/repo.git`
2. `httpSubModuleUrl` has the URL with the username: `https://oauth2@mygitlab.com/path/to/my/repo.git`

Then, renovate is using `https://oauth2@mygitlab.com/path/to/my/repo.git` for the `git ls-remote` command.
Since the environment variable is set to match `https://mygitlab.com` (and **not ``*https://oauth2@mygitlab.com`**), `git ls-remote` asks for the password.

Looking at `util/git/url.ts`, it seems like the parsed URL has 3 properties to manage state and makes some confusion on that scenario:
https://github.com/renovatebot/renovate/blob/4b616c29c31b3cf891c1244fabae944cea9efb65/lib/util/git/url.ts#L11-L12

We should be able to confirm that by adding the following tests [here](https://github.com/renovatebot/renovate/blob/main/lib/util/git/url.spec.ts#L38):

```js
it('removes username/password from URL', () => {
    expect(getHttpUrl('https://user:password@foo.bar/')).toBe('https://foo.bar/');
});

it('replaces username/password with given token', () => {
    expect(getHttpUrl('https://user:password@foo.bar/', 'another-user:a-secret-pwd')).toBe('https://another-user:a-secret-pwd@foo.bar/');
});
```
The fix to make it work, then would be clearing state [here](https://github.com/renovatebot/renovate/blob/main/lib/util/git/url.ts#L18):

```diff
  parsedUrl.user = null;
  parsedUrl.password = null;
  parsedUrl.token = token ?? '';
```

**Reply by lcobucci** [reply]
~What I found interesting is that `git-url-parse` doesn't have the `password` property in its source, however it does test for it:~

~https://github.com/IonicaBizau/git-url-parse/blob/4f0dd57c822a220a13913a9aaabfb0cb59051a30/test/index.js#L105-L113~

Never mind, it comes from [here](https://github.com/IonicaBizau/parse-path/blob/master/lib/index.js#L52).

**Reply by lcobucci** [reply]
@Shegox https://github.com/renovatebot/renovate/pull/24306 should do the trick :+1:

**Reply by Shegox** [reply]
That's great. I was missing the point that it was a relative URL, that explains a lot of things :)

**Reply by lcobucci** [reply]
> That's great. I was missing the point that it was a relative URL, that explains a lot of things :) 

Apologies for forgetting that detail! I'm juggling with way too many things at the moment...

**Reply by lcobucci** [reply, marked as answer]
@Shegox just ran jobs with the latest release and can confirm that our fixes addressed the issue.
Thanks a lot for your support.
